File: backend/src/main/java/finalproject/backend/service/LessonProgressService.java
package finalproject.backend.service;

import finalproject.backend.request.LessonProgressRequest;
import finalproject.backend.response.ApiResponse;
import finalproject.backend.response.LessonProgressResponse;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface LessonProgressService {

    /** Create or update progress for a lesson. */
    ApiResponse<LessonProgressResponse> upsertProgress(LessonProgressRequest request);

    /** Mark a lesson as 100 % completed for the given user. */
    ApiResponse<LessonProgressResponse> markCompleted(Long userId, Long lessonId);

    /** Get progress for a specific user + lesson pair. */
    ApiResponse<LessonProgressResponse> getProgress(Long userId, Long lessonId);

    /** Get all lesson progress records for a user. */
    ApiResponse<List<LessonProgressResponse>> getProgressByUser(Long userId);

    /** Get all lesson progress records for a user within a specific course. */
    ApiResponse<List<LessonProgressResponse>> getProgressByCourseAndUser(Long courseId, Long userId);

    /** Count how many lessons a user has completed (globally). */
    ApiResponse<Long> countCompletedByUser(Long userId);

    /** Count how many lessons a user has completed in a specific course. */
    ApiResponse<Long> countCompletedByCourseAndUser(Long courseId, Long userId);

    /** Delete a progress record (admin or owner only). */
    ApiResponse<Void> deleteProgress(Long userId, Long lessonId);
}

File: frontend/src/services/lessonProgressService.ts
import { get, post, del } from '../lib/BaseApi'
import type {
  LessonProgressResponse,
  LessonProgressRequest,
} from '../types/lessonProgressType'


// ═════════════════════════════════════════════════════════════
//  8. LESSON PROGRESS SERVICE
// ══════════════════════════════════════════════════

const PROGRESS_PATH = '/api/v1/lesson-progress'

export const lessonProgressService = {
  /** POST / — upsert scroll position, reading time, completion */
  upsert: (payload: LessonProgressRequest): Promise<LessonProgressResponse> => {
    return post<LessonProgressResponse>(PROGRESS_PATH, payload)
  },

  /** POST /complete?lessonId= — mark lesson completed */
  markCompleted: (lessonId: number): Promise<LessonProgressResponse> => {
    return post<LessonProgressResponse>(`${PROGRESS_PATH}/complete`, null, { params: { lessonId } })
  },

  /** DELETE /?lessonId= — remove progress record */
  remove: (lessonId: number): Promise<void> => {
    return del<void>(PROGRESS_PATH, { params: { lessonId }, raw: true })
  },

  /** GET /?lessonId= — single lesson progress */
  get: (lessonId: number): Promise<LessonProgressResponse> => {
    return get<LessonProgressResponse>(PROGRESS_PATH, { params: { lessonId } })
  },

  /** GET /me — all progress for authenticated user */
  getMine: (): Promise<LessonProgressResponse[]> => {
    return get<LessonProgressResponse[]>(`${PROGRESS_PATH}/me`)
  },

  /** GET /course/:courseId — progress within a course */
  getByCourse: (courseId: number): Promise<LessonProgressResponse[]> => {
    return get<LessonProgressResponse[]>(`${PROGRESS_PATH}/course/${courseId}`)
  },

  /** GET /me/completed-count — total completed lessons */
  countCompleted: (): Promise<number> => {
    return get<number>(`${PROGRESS_PATH}/me/completed-count`)
  },

  /** GET /course/:courseId/completed-count */
  countCompletedByCourse: (courseId: number): Promise<number> => {
    return get<number>(`${PROGRESS_PATH}/course/${courseId}/completed-count`)
  }
}

//